flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java

```diff
@@ -0,0 +1,354 @@
+
+			public UUID rmLeaderSessionId;
+
+			MockResourceManagerRuntimeServices() throws Exception {
+				scheduledExecutor = mock(ScheduledExecutor.class);
+				highAvailabilityServices = new TestingHighAvailabilityServices();
+				rmLeaderElectionService = new TestingLeaderElectionService();
+				highAvailabilityServices.setResourceManagerLeaderElectionService(rmLeaderElectionService);
+				heartbeatServices = new TestingHeartbeatServices(5L, 5L, scheduledExecutor);
+				metricRegistry = mock(MetricRegistry.class);
+				slotManager = new SlotManager(
+						new ScheduledExecutorServiceAdapter(new DirectScheduledExecutorService()),
+						Time.seconds(10), Time.seconds(10), Time.minutes(1));
+				jobLeaderIdService = new JobLeaderIdService(
+						highAvailabilityServices,
+						rpcService.getScheduledExecutor(),
+						Time.minutes(5L));
+			}
+
+			public void grantLeadership() throws Exception {
+				rmLeaderSessionId = UUID.randomUUID();
+				rmLeaderElectionService.isLeader(rmLeaderSessionId).get(timeout.toMilliseconds(), TimeUnit.MILLISECONDS);
+			}
+		}
+
+		/**
+		 * Start the resource manager and grant leadership to it.
+		 */
+		public void startResourceManager() throws Exception {
+			resourceManager.start();
+			rmServices.grantLeadership();
+		}
+
+		/**
+		 * Stop the Akka actor system.
+		 */
+		public void stopResourceManager() throws Exception {
+			rpcService.stopService();
+		}
+	}
+
+	static class TestContainer extends UtilsTest.TestingContainer {
+		Resource resource;
+		Priority priority;
+
+		TestContainer(String id, String host) {
+			super(id, host);
+		}
+
+		@Override
+		public Resource getResource() {
+			return resource;
+		}
+
+		@Override
+		public void setResource(Resource resource) {
+			this.resource = resource;
+		}
+
+		@Override
+		public Priority getPriority() {
+			return priority;
+		}
+
+		@Override
+		public void setPriority(Priority priority) {
+			this.priority = priority;
+		}
+	}
+
+	@Test
+	public void testStopWorker() throws Exception {
+		new Context() {{
+			startResourceManager();
+			// Request slot from SlotManager.
+			resourceManager.runInMainThread(() -> {
+				try {
+					rmServices.slotManager.registerSlotRequest(
+							new SlotRequest(new JobID(), new AllocationID(), resourceProfile1, taskHost));
+				} catch (SlotManagerException e) {
+					e.printStackTrace();
+				}
+			});
+
+			// Callback from YARN when container is allocated.
+			Container testingContainer = new TestContainer(task.toString(), taskHost);
+			testingContainer.setResource(Resource.newInstance(200, 1));
+			testingContainer.setPriority(Priority.UNDEFINED);
+			resourceManager.onContainersAllocated(ImmutableList.of(testingContainer));
+
+			// Remote task executor registers with YarnResourceManager.
+			TaskExecutorGateway mockTaskExecutorGateway = mock(TaskExecutorGateway.class);
+			rpcService.registerGateway(taskHost, mockTaskExecutorGateway);
+			resourceManager.registerTaskExecutor(taskHost,
```

Comment:
You shouldn't directly interact with the `ResourceManager` but instead use the `ResourceManagerGateway` when calling RPCs. Otherwise it is not guaranteed that you won't have concurrent operations (one by the test thread and the other by the RM's main thread).